packages/discovery-provider/src/api/v1/models/purchasable_content.py

```diff
@@ -0,0 +1,134 @@
+from flask_restx import fields
+from flask_restx.fields import MarshallingError, marshal
+
+from .common import ns
+from .playlists import (
+    full_playlist_model,
+    full_playlist_without_tracks_model,
+    playlist_model,
+)
+from .tracks import track, track_full
+
+
+class PurchasableContentItem(fields.Raw):
+    def format(self, value):
+        try:
+            if value.get("track_id"):
+                return marshal(value, track)
+            if value.get("playlist_id"):
+                return marshal(value, playlist_model)
+        except Exception as e:
+            raise MarshallingError(
+                "Unable to marshal as purchasable_content item"
+            ) from e
+
+
+class PurchasableContentItemFull(fields.Raw):
+    def format(self, value):
+        try:
+            if value.get("track_id"):
+                return marshal(value, track_full)
+            if value.get("playlist_id"):
+                return marshal(value, full_playlist_model)
+        except Exception as e:
+            raise MarshallingError(
+                "Unable to marshal as purchasable_content item"
+            ) from e
+
+
+purchasable_content_model = ns.model(
+    "purchasable_content",
+    {
+        "content_type": fields.String(enum=["track", "album"], required=True),
+        "item": PurchasableContentItem(required=True),
+        "class": fields.String(required=True, discriminator=True),
+    },
+)
+
+track_purchasable_content_model = ns.inherit(
+    "track_purchasable_content",
+    purchasable_content_model,
+    {
+        "content_type": fields.String(enum=["track"], required=True),
+        "item": fields.Nested(track, required=True),
+    },
+)
+
+collection_purchasable_content_model = ns.inherit(
+    "collection_purchasable_content",
+    purchasable_content_model,
+    {
+        "content_type": fields.String(enum=["album"], required=True),
+        "item": fields.Nested(playlist_model, required=True),
+    },
+)
+
+purchasable_content_full_model = ns.model(
+    "purchasable_content_full",
+    {
+        "content_type": fields.String(enum=["track", "album"], required=True),
+        "item": PurchasableContentItemFull(required=True),
+        "class": fields.String(required=True, discriminator=True),
+    },
+)
+
+track_purchasable_content_full_model = ns.inherit(
+    "track_purchasable_content_full",
+    purchasable_content_full_model,
+    {
+        "content_type": fields.String(enum=["track"], required=True),
+        "item": fields.Nested(track_full, required=True),
+    },
+)
+
+collection_purchasable_content_full_model = ns.inherit(
+    "collection_purchasable_content_full",
+    purchasable_content_full_model,
+    {
+        "content_type": fields.String(enum=["album"], required=True),
+        "item": fields.Nested(full_playlist_model, required=True),
+    },
+)
+
+collection_purchasable_content_full_without_tracks_model = ns.inherit(
```

Comment:
+1, the polymorph pattern was a thing I whipped up because our OneOf hack didn't do what I wanted yet. But I think Marcus' new implementation with discriminator is probably easier to work with and understand.

Reply:
Additionally: If you copied this code from reposts, this model probably isn't used. In reposts, it was a one-off used for endpoints that returned a list of only playlist reposts.